src/main/java/org/repodriller/scm/GitRepository.java

```diff
@@ -51,33 +51,45 @@
 import org.repodriller.domain.Commit;
 import org.repodriller.domain.Developer;
 import org.repodriller.domain.ModificationType;
-import org.repodriller.util.FileUtils;
+import org.repodriller.util.RDFileUtils;
 
 public class GitRepository implements SCM {
 
+	/* Constants. */
 	private static final int MAX_SIZE_OF_A_DIFF = 100000;
 	private static final int DEFAULT_MAX_NUMBER_OF_FILES_IN_A_COMMIT = 5000;
 	private static final String BRANCH_MM = "mm";
 
-	private String path;
-	private String mainBranchName;
-	private int maxNumberFilesInACommit;
-	private int maxSizeOfDiff;
+	/* Auto-determined. */
+	private String mainBranchName = null;
+	private int maxNumberFilesInACommit = -1;
+	private int maxSizeOfDiff = -1;
 
 	private static Logger log = Logger.getLogger(GitRepository.class);
-	private boolean firstParentOnly;
 
-	public GitRepository(String path, boolean firstParentOnly) {
-		this.path = path;
-		this.firstParentOnly = firstParentOnly;
-		this.maxNumberFilesInACommit = checkMaxNumberOfFiles();
-		this.maxSizeOfDiff = checkMaxSizeOfDiff();
+	/* User-specified. */
+	private String path = null;
+	private boolean firstParentOnly = false;
+
+	/**
+	 * Make sub-classing easier. Make sure you initialize appropriately with the Setters.
+	 */
+	public GitRepository() {
```

Comment:
not sure if i like this constructor here. It instantiates the class in an invalid state. I'd remove the default constructor, and write one for the child class if you think you need it.

Reply:
So this and all of your subsequent comments about getters and setters all have the same root cause: In this PR, GitRemoteRepository extends GitRepository.

We need to call the super constructor somehow, and the super constructor must be called "first thing" in any sub-class's constructor. But at that point in GitRemoteRepository we don't have a path because we haven't cloned the remote repository to a local temp directory.

All of the existing GitRepository constructors require a path, so I added another constructor that doesn't require a path and added setters so the GitRemoteRepository could finish the initialization by itself.

Thoughts? Other ideas?

I think having GitRemoteRepository extend GitRepository is a win, but I'm not sure how to get the interface clean.

Reply:
Hmm, maybe this constructor can be `protected` rather than `public` then. At least, only children (and classes on the same package) can access.

Reply:
Addressed in a03dbd8